test/extended/oauth/requestheaders.go

```diff
@@ -0,0 +1,390 @@
+
+	exutil "github.com/openshift/origin/test/extended/util"
+)
+
+func init() {
+	utilruntime.Must(osinv1.Install(scheme))
+}
+
+const clientCAName = "test-client-ca"
+
+type certAuthTest struct {
+	name          string
+	cert          *x509.Certificate
+	key           *rsa.PrivateKey
+	endpoint      string
+	expectToken   bool
+	expectedError string
+}
+
+// helper error to retrieve the token from a redirect chain stopped by custom
+// http client CheckRedirect
+type tokenFoundError struct{}
+
+func (e *tokenFoundError) Error() string {
+	return fmt.Sprintf("token found")
+}
+
+var _ = g.Describe("[Serial] [Feature:OAuthServer] [RequestHeaders] [IdP]", func() {
+	var oc = exutil.NewCLI("request-headers", exutil.KubeConfigPath())
+
+	g.It("test RequestHeaders IdP", func() {
+		caCert, caKey := createClientCA(oc.AdminKubeClient().CoreV1())
+		defer oc.AdminKubeClient().CoreV1().ConfigMaps("openshift-config").Delete(clientCAName, &metav1.DeleteOptions{})
+
+		changeTime := time.Now()
+		oauthCluster, err := oc.AdminConfigClient().ConfigV1().OAuths().Get("cluster", metav1.GetOptions{})
+		o.Expect(err).NotTo(o.HaveOccurred())
+		oauthCluster.Spec.IdentityProviders = []configv1.IdentityProvider{
+			{
+				Name: "test-request-header",
+				IdentityProviderConfig: configv1.IdentityProviderConfig{
+					Type: configv1.IdentityProviderTypeRequestHeader,
+					RequestHeader: &configv1.RequestHeaderIdentityProvider{
+						ClientCA: configv1.ConfigMapNameReference{
+							Name: clientCAName,
+						},
+						Headers:                  []string{"X-Remote-User"},
+						EmailHeaders:             []string{},
+						NameHeaders:              []string{},
+						PreferredUsernameHeaders: []string{},
+						LoginURL:                 "https://dontcare.com/web-login/oauth/authorize?${query}",
+						ChallengeURL:             "https://dontcare.com/challenges/oauth/authorize?${query}",
+					},
+				},
+			},
+		}
+		_, err = oc.AdminConfigClient().ConfigV1().OAuths().Update(oauthCluster)
+		o.Expect(err).NotTo(o.HaveOccurred())
+		// clean up after ourselves
+		defer func() {
+			oauthCluster, _ := oc.AdminConfigClient().ConfigV1().OAuths().Get("cluster", metav1.GetOptions{})
+			oauthCluster.Spec = configv1.OAuthSpec{}
+			oc.AdminConfigClient().ConfigV1().OAuths().Update(oauthCluster)
+		}()
+
+		oauthURL := getOAuthServerURL(oc)
+		goodCert, goodKey := generateCert(caCert, caKey, []x509.ExtKeyUsage{x509.ExtKeyUsageClientAuth})
+
+		caCert2, caKey2 := generateCA("The Other Testing CA")
+		badCert, badKey := generateCert(caCert2, caKey2, nil)
+
+		testCases := []certAuthTest{
+			{
+				name:          "/authorize: anonymous",
+				endpoint:      "/oauth/authorize?client_id=openshift-challenging-client&response_type=token",
+				expectedError: "302 Found",
+			},
+			{
+				name:        "/authorize: valid cert",
+				cert:        goodCert,
+				key:         goodKey,
+				endpoint:    "/oauth/authorize?client_id=openshift-challenging-client&response_type=token",
```

Comment:
You need at least some tests for a browser based client id, i.e. not `openshift-challenging-client`

Reply:
Added two + /oauth/token/request, a bit too simple, but it's something